### ISSUE TITLE
Persistent connection support for `raw` sockets (ngx.req.socket(true))

### ISSUE THREAD

**Issue by piersoh** [openresty/lua-nginx-module#1527]
It seems that OpenResty doesn't support persistent connections when handling requests using `ngx.req.socket(true)`. If one runs the code from test#15:
https://github.com/openresty/lua-nginx-module/blob/master/t/116-raw-req-socket.t#L785
Note: There is a bug with this test as `err` is used as both a variable and function name so one needs to be changed (e.g. `sed 's/err(/errfn(/g' 116-raw-req-socket.t`) 

It can be seen that if a client sends more than one request (with header Connection: keep-alive) down the same TCP connection only the first one is acted on. It works fine if the client closes the connection and opens a new each time but this is inefficient. Also lua scripts can clearly handle persistent connections when using non-raw socket `ngx.req.socket()`.  I'm wondering if there's anything missing from that test code to support persistent connections or if OpenResty (tested v1.13.6.2 & v1.15.8.1) doesn't support it?  Support for it would be good.

**Comment by agentzh** [comment]
@piersoh Nope, raw req sockets are protocol agnostic so it is not safe to reuse them for future connections (note that the next request handler may expect it to be proper HTTP request and there is no way to enforce that for raw req sockets and will lead to very hard-to-debug problems). This is intentional.

**Comment by piersoh** [comment]
@agentzh Thanks for the response - would it not be possible to add an explicit function that could be called to signal the end of that session - in a way that could be equivalent to a client connection termination - but to keep the socket open? I see there might be issues if data is left in the buffers, but doesn't the nginx parser just drop any malformed requests?

Otherwise it is not clear how to handle streamed persistent chunked mode requests (i.e. not using nginx's chunked request handling), with OpenResty?

However if such functionality were added it would be good to have enough control over the flow so one could handle chunks in lua, and potentially extract/handle such things as chunk extensions and trailers etc.

**Comment by agentzh** [comment]
@piersoh 

> but doesn't the nginx parser just drop any malformed requests?

Not really. Nginx does support HTTP pipelined requests so unread data in a connection is very normal protocol wise.

> Otherwise it is not clear how to handle streamed persistent chunked mode requests (i.e. not using nginx's chunked request handling), with OpenResty?

You just handle any more requests directly in the same request handler. Otherwise it's a wrong use case for using the http lua module. Maybe you should use ngx stream lua module for everything yourself in Lua?

**Comment by piersoh** [comment]
@agentzh 

> Not really. Nginx does support HTTP pipelined requests so unread data in a connection is very normal protocol wise.
Ok - Sorry I didn't mean that Nginx would drop the connection but as you say it can deal with unread data so if it was possible to explicitly terminate the request in lua then Nginx could potentially handle any unread data?

> You just handle any more requests directly in the same request handler. Otherwise it's a wrong use case for using the http lua module.
Ah ok - I guess I could take that approach but it feels strange to do it in that way as I'd have to re-implement request path and header handling, when all I want to do is just handle the chunked request body - which I can do now but only if the client terminates each connection.

>  Maybe you should use ngx stream lua module for everything yourself in Lua?
Maybe but it seems like overkill using the streaming module for just handling persistent HTTP chunked requests at the chunk level - It's not like it is a websocket or some other custom control channel that I'm trying to use here...
